app/models/mapping.rb

```diff
@@ -221,9 +230,60 @@ def dataset_path_must_exist
     errors.add(:base, "Dataset must exist.") unless dataset.exist?
   end
 
+  # Validator for app
+  def app_path_may_not_be_blank
+    errors.add(:base, "You must select an app.") unless app.exist?
+  end
+
   # Directories between $HOME and the configured database directory must be set to 775
   # @return [Boolean]
   def self.directory_perms_are_775?(directory)
     directory.stat.mode.to_s(8).end_with?('775')
   end
+
+  # Checks to see if FACLs are modifiable by the user
+  #
+  # Note that this check assumes that no negative permissions have been set.
+  #
+  # @return [Boolean]
+  def self.can_modify_facl?(pathname)
+    pathname.exist? && ( pathname.owned? || group_facl_entry_has_C_set?(pathname) )
+  end
+
+  # Get the group owner's name
+  # @return [String]
+  def self.get_groupname_for_pathname(pathname)
+    Etc.getgrgid(pathname.stat.gid).name
+  end
+
+  # Check if pathname is owned by the admin_group
+  # @return [Boolean]
+  def self.app_has_correct_group_ownership?(pathname)
+    get_groupname_for_pathname(pathname) == Configuration.admin_group
+  end
+
+  # Does the app have permission modification enabled for the GROUP principle
+  # @return [Boolean]
+  def self.group_facl_entry_has_C_set?(pathname)
+    result = OodSupport::ACLs::Nfs4ACL.get_facl(
+      path: pathname
+    ).entries.select{
+      |entry| entry.principle == 'GROUP'
+    }.first.permissions.include?(:C)
+
+    logger.debug "group_facl_entry_has_C_set?(#{pathname}) == #{result}"
+
+    result
+  end
+
+  def self.get_mapping_for_id(id)
+    begin
+      return Mapping.find(id)
+    rescue ActiveRecord::RecordNotFound
+      OpenStruct.new(
+        :destroy_and_remove_facls => false,
+        :save_message => "No mapping exists with id #{id}"
+      )
```

Comment:
Instead of having this, why can't we just do in the controller:

```
mapping = Mapping.find(params[:id])
```

and in the controller handle the `ActiveRecord::RecordNotFound`. Is it necessary to handle the `ActiveRecord::RecordNotFound`, which results in a 404?

Reply:
One situation where they would see it is if there is a delay between submitting the first delete and immediately sending another. I know quite a few people who click buttons more than once.

I agree that the OpenStruct usage isn't ideal, it's just something that I was playing with on Friday. I liked that it let me avoid having to add an error handler for RecordNotFound.

Reply:
As far as the multiple button clicks, the presence of the alert should prevent this from occurring: https://github.com/OSC/shiny_admin/blob/46261c7d18decbc60f9be84ec67855ef3baecc19/app/views/mappings/index.html.erb#L27

A single click presents the confirmation alert, and only if that is "okay" does the delete request get sent.

Reply:
And of course as we discussed offline, I think its appropriate to not add rescue for RecordNotFound since Rails should handle this to be 404.